Describe how to specify the display specs when running `amc benchmark` (#12)
anbox-cloud/discourse-docs: benchmarking-a-deployment.md

Anbox Cloud provides different tools to allow benchmarking different aspects of a deployment. This pages describes the available tools and how to use them.

## Container Benchmarking

The `amc` command line utility comes with a subcommand to allow easy benchmarking of an Anbox Cloud deployment. It can be used to measure the time containers take to start up but also for measuring their frames per second. The benchmark enables you to put Anbox Cloud under load. The results can be used to evaluate the performance of Anbox Cloud for a well defined workload.

> **Warning:** If your application is not constantly refreshing the screen by itself, like regular Android applications providing a simple user interface do, you will get a low FPS number. The benchmark can only provide useful information if you know your workload and how it should perform in an ideal scenario.

### Run the Benchmark

Have a look at the output of

```bash

> $ amc benchmark -h

Reviewer: Yes, we have that as a pattern elsewhere so should do the same here

> $ amc benchmark -h

Reviewer 2: I'll create an issue to change that everywhere.

2019/01/21 11:15:25 Removing container bh2qijvo3v1lt1ft4nlg ...
2019/01/21 11:15:25 Removing container bh2qimfo3v1lt1ft4nmg ...
2019/01/21 11:15:25 Removing container bh2qiovo3v1lt1ft4nng ...
2019/01/21 11:15:25 Removing container bh2qi4vo3v1lt1ft4nfg ...
2019/01/21 11:15:25 Removing container bh2qirfo3v1lt1ft4nog ...
2019/01/21 11:15:25 Removing container bh2qhvvo3v1lt1ft4ndg ...
2019/01/21 11:15:25 Removing container bh2qi9vo3v1lt1ft4nhg ...
2019/01/21 11:15:39 Containers boot time measurement:
2019/01/21 11:15:39   Launching all 15 containers took 2m36.560310342s
2019/01/21 11:15:39   Out of 15 containers 0 failed to launch
2019/01/21 11:15:39   Average container launch time: 6.149119411s
2019/01/21 11:15:39   Max container launch time: 6.576302043s
2019/01/21 11:15:39   Min container launch time: 5.911184959s
2019/01/21 11:15:39   Android system failed to boot in the following containers:
2019/01/21 11:15:39     None
2019/01/21 11:15:39 Containers statistics:
2019/01/21 11:15:39   FPS avg 58 min 52 max 64 for 15 containers
2019/01/21 11:15:39   Containers below run at low FPS(<30):
2019/01/21 11:15:39     None
```

> The benchmark command launches 15 containers on the Anbox `null` platform with the following default display specification:

Reviewer 3: ```suggestion
The benchmark command launches 15 containers on the Anbox [`null` platform](https://anbox-cloud.io/docs/reference/anbox-platforms) with the following default display specification:
```

Author: Fixed.

 | Display Spec            | Value |
 | ----------------------- | ----- |
 | Display width           | 1280  |
 | Display height          | 720   |
 | FPS                     | 60    |
 | Display density         | 160   |

You can configure a different display specification through the `--user-data` parameter when running the benchmark. The required format for the parameter varies based on the different platforms:

| Platform              | Required format of user data          | Example                                                   |
| --------------------- | ------------------------------------  | --------------------------------------------------------  |
| Null                  | Comma-separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |
| Swrast                | Comma-separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |

> | WebRTC                | Json-based                            | {<br>"display_width": <display_width>,<br>"display_height": <display_height>,<br>"display_density": <display_density>,<br>"fps": <display_fps>,<br>"render_only": true<br> } |

Reviewer 3: Let's remove those `<br>`, they are probably a result of some bad templating

Author: Per discussion in MP,  this is wrong.

The application is a stress test which is primarily being used to verify the OpenGL emulation of the Android emulator but serves general stress test purposes as well.

The application was modified to constantly run the benchmark scene with `~1000` objects to drive CPU and GPU utilization of a container. You can find the application APK package here:

* [Stress Test Application (amd64 + arm64, sha256: dfb67bb97ceb5e3c64398210e2ecd65286f14b0c67856abde2ac0bead07bd223)](https://oem-share.canonical.com/partners/indore/share/releases/1.8/other/com.android.gpu_emulation_stress_test_1.8.apk)

Once downloaded you can add the application as regular (see [Application Management](https://discourse.ubuntu.com/t/managing-applications/17760)) to your Anbox Cloud installation. The manifest may look like:

```yaml
name: gpu-stress-test
instance: a2.3
image: default
```

> **Hint:** Depending on the selected instance type and how powerful a single core of your machines is you may get better results with a larger instance type.

## Stream Benchmarking

> As streaming involves more things to automate for a proper benchmark Anbox Cloud provides a dedicated benchmark tool which allows creating a streaming session, receiving the video/audio stream and collecting various statistics and optional also dumping the received stream to a local file.

Reviewer 3: ```suggestion
As streaming involves more things to automate for a proper benchmark, Anbox Cloud provides a dedicated benchmark tool which allows creating a streaming session, receiving the video/audio stream and collecting various statistics and optional also dumping the received stream to a local file.
```

> **WARNING:** Right now the benchmark tool is only supported on a Linux system supporting snaps and on 64 bit x86 systems. Support for 64 bit ARM systems will be added at a later point.

The benchmark tool comes with the `anbox-cloud-tests` snap which you can install with

```bash

> $ snap install anbox-cloud-tests

Reviewer 3: Same as above, let's remove the `$`

```

Once the snap is installed you can have a look at the supported command line options of the benchmark tool:

```bash

> $ anbox-cloud-tests.benchmark -h

Reviewer 3: same here

```

To run the benchmark against an existing Anbox Cloud deployment, use the following as an example but check the other available options the benchmark offers too:

```bash

> $ anbox-cloud-tests.benchmark \

Reviewer 3: and here too :)

2019/01/21 11:15:25 Removing container bh2qijvo3v1lt1ft4nlg ...
2019/01/21 11:15:25 Removing container bh2qimfo3v1lt1ft4nmg ...
2019/01/21 11:15:25 Removing container bh2qiovo3v1lt1ft4nng ...
2019/01/21 11:15:25 Removing container bh2qi4vo3v1lt1ft4nfg ...
2019/01/21 11:15:25 Removing container bh2qirfo3v1lt1ft4nog ...
2019/01/21 11:15:25 Removing container bh2qhvvo3v1lt1ft4ndg ...
2019/01/21 11:15:25 Removing container bh2qi9vo3v1lt1ft4nhg ...
2019/01/21 11:15:39 Containers boot time measurement:
2019/01/21 11:15:39   Launching all 15 containers took 2m36.560310342s
2019/01/21 11:15:39   Out of 15 containers 0 failed to launch
2019/01/21 11:15:39   Average container launch time: 6.149119411s
2019/01/21 11:15:39   Max container launch time: 6.576302043s
2019/01/21 11:15:39   Min container launch time: 5.911184959s
2019/01/21 11:15:39   Android system failed to boot in the following containers:
2019/01/21 11:15:39     None
2019/01/21 11:15:39 Containers statistics:
2019/01/21 11:15:39   FPS avg 58 min 52 max 64 for 15 containers
2019/01/21 11:15:39   Containers below run at low FPS(<30):
2019/01/21 11:15:39     None
```

> This launches 15 containers on Anbox null platform with default display specification

Reviewer 2: ```suggestion
The benchmark command launches 15 containers on the Anbox `null` platform with the following default display specification:
```

Author: Will fix

 | Display Spec            | Value |
 | ----------------------- | ----- |
 | Display width           | 1280  |
 | Display height          | 720   |
 | FPS                     | 60    |
 | Display density         | 160   |

> for benchmarking.

Reviewer 2: ```suggestion
```

Author: Will fix

> A different display specification can be configured through `--user-data` parameter when running the benchmark. The content format of user-data varies based on different platforms.

Reviewer 2: ```suggestion
You can configure a different display specification through the `--user-data` parameter when running the benchmark. The required format for the parameter varies based on the different platforms:
```

Author: Will fix

| Platform              | Content format of user data           | Example                                                   |
| --------------------- | ------------------------------------  | --------------------------------------------------------  |

> | Null                  | Comma separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |
> | Swrast                | Comma separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |

Reviewer 2: ```suggestion
| Null                  | Comma-separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |
| Swrast                | Comma-separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |
```

Author: Will fix

| WebRTC                | Json-based                            | {<br>"display_width": <display_width>,<br>"display_height": <display_height>,<br>"display_density": <display_density>,<br>"fps": <display_fps>,<br>"render_only": true<br> } |

> > **Note:** when running benchmark against webrtc platform, Anbox session must be launched with render only mode since `amc benchmark` doesn't require WebRTC negotiation to establish WebRTC connection upon container creation.

Reviewer 2: ```suggestion
> **Note:** When running a benchmark against the `webrtc` platform, you must launch the Anbox session in render-only mode. The `amc benchmark` command doesn't require WebRTC negotiation to establish a WebRTC connection upon container creation.
```
Is it correct that the user launches the session?
And is it really a "must"? The explanation sounds to me like render-only mode is sufficient, but you could also use other modes.

Author: Is it correct that the user launches the session?
A:  Whenever creating a container, internally we launch a fresh Anbox session, which talks to Anbox container manager to boot up Android system. Perhaps to make it easy to understand, we may stick to **launch a container** ?
      
And is it really a "must"? The explanation sounds to me like render-only mode is sufficient, but you could also use other modes.
A: yes, it is. otherwise, creating containers would fail immediately right after they just got started due to the missing session URL since the `amc benchmark` command doesn't interact with stream gateway, hence no session URL is given at all.  

I am not quite sure if we want to leak too many implementation details about this. What about 

> **Note:** Because the `amc benchmark` command doesn't interact with stream gateway for the benchmark execution, when running a benchmark against the `webrtc` platform, you must launch containers in render-only mode, otherwise container creation would fail during the benchmark runtime.

Reviewer 2: But then it is the benchmark command that launches the containers, and not the user (at least not directly)?

Maybe:
```suggestion
> **Note:** Because the `amc benchmark` command doesn't interact with the stream gateway for the benchmark execution, the containers are launched in render-only mode when running a benchmark against the `webrtc` platform. Otherwise, the container creation would fail.
```

Author: But then it is the benchmark command that launches the containers, and not the user (at least not directly)?
A: yes, it's the benchmark command that launches the containers, however the display spec is specified by a user, not benchmark command itself. We make it clear that people have to declare the `render-only: true` in the display specs against webrtc platform.  See the json-based content in the display spec table. Things can easily mess up if `render-only: true` is not present in the display spec, which is going to be specified by end user.

Reviewer 2: Ah, right - now I get it. I missed the render-only in the table. ;)

Then how about:
```suggestion
> > **Note:** If you're running a benchmark against the `webrtc` platform, make sure to specify `"render_only": true` to launch the containers in render-only mode. Otherwise, the container creation will fail, because the `amc benchmark` command doesn't interact with the stream gateway for the benchmark execution.
```

Author: +1, much better.
